### DISCUSSION TITLE
Luatos C3 display (Air101-LCD)

### DISCUSSION THREAD

**Discussion by Jason2866** [arendst/Tasmota#18350] [Q&A] [answered]
Struggling with display.ini for the C3 Luatos ST7735S Display.
![Luatos_display](https://user-images.githubusercontent.com/24528715/230410963-636d2fce-4cd9-4539-a9dc-c709fe3862a3.jpg)

This demo code works fine. https://github.com/MoonFox2006/Air101-LCD
Managed to get display activated using this template and display.ini (wrong init sequences).

```
{"NAME":"LuatosC3","GPIO":[1,6210,736,704,1,1,800,768,1,1,8736,0,544,1,0,0,0,0,1,1,1,1],"FLAG":0,"BASE":1}
```

```
:H,ST7735S,160,80,16,SPI,1,7,2,3,6,11,10,*,16
:S,2,1,1,0,30,30
:I
01,80
11,80
B1,3,01,2C,2D
B2,3,01,2C,2D
B3,6,01,2C,2D,01,2C,2D
B4,1,07
C0,3,A2,04,84
C1,1,C5
C2,2,0A,00
C3,2,8A,2A
C4,2,8A,EE
C5,1,0E
20,0
36,1,C8
3A,1,05
2A,4,00,02,00,7F
2B,4,00,01,00,7F
E0,10,02,1C,07,12,37,32,29,2D,29,25,2B,39,00,01,03,10
E1,10,03,1D,07,06,2E,2C,29,2D,2E,2E,37,3F,00,00,02,10
13,80
29,80
:o,28
:O,29
:A,2A,2B,2C,16
:R,36
:0,60,00,00,00
:i,20,21
:B,60,1
#
```
![Luatos_Display](https://user-images.githubusercontent.com/24528715/230411450-423a50db-3837-413f-b04c-8130715e1a84.png)
![image](https://user-images.githubusercontent.com/24528715/230412834-c75c7f02-d94a-4ff5-9711-edb893886dcf.png)

**Comment by Jason2866** [comment]
@gemu2015 Can you help?

**Reply by gemu2015** [reply, marked as answer]
ok try to change this line in descriptor
:0,60,00,00,00
to
:0,60,00,18,00

**Comment by Jason2866** [comment]
@gemu2015 Thx! Works now :-)
![IMG_20230424_164435](https://user-images.githubusercontent.com/24528715/234033324-937885c5-67cd-4429-8d18-e1e8895c58bf.jpg)
The working display.ini
```
:H,ST7735S,160,80,16,SPI,1,7,2,3,6,11,10,*,16
:S,2,1,1,0,30,30
:I
01,80
11,80
B1,3,01,2C,2D
B2,3,01,2C,2D
B3,6,01,2C,2D,01,2C,2D
B4,1,07
C0,3,A2,04,84
C1,1,C5
C2,2,0A,00
C3,2,8A,2A
C4,2,8A,EE
C5,1,0E
20,0
36,1,C8
3A,1,05
2A,4,00,02,00,7F
2B,4,00,01,00,7F
E0,10,02,1C,07,12,37,32,29,2D,29,25,2B,39,00,01,03,10
E1,10,03,1D,07,06,2E,2C,29,2D,2E,2E,37,3F,00,00,02,10
13,80
29,80
:o,28
:O,29
:A,2A,2B,2C,16
:R,36
:0,60,00,18,00
:i,20,21
:B,60,1
#
```

**Comment by barbudor** [comment]
Hi @Jason2866 
Just started mine uisng your display.ini
I haven't been far for now but I've noticed some random pixels on the 1 or 2 bottom lines as well as 1 or to right columns.
Do you see the same ?

**Comment by Jason2866** [comment]
No, I don't have ghost pixels. Tried reducing Display clock?

**Comment by booroondook-eng** [comment]
Hi all! I have Luatos ESP32C3 board and Air101-LCD display.
What firmware should I use for this board? I was looking for "...32c3-display..." and "...32c3-lvgl..." firmware, but I couldn't find anything.
I also wanted to compile the firmware by myself using PlatformIO, but I don't know what files need to be edited to change the configuration to support such a display.

**Comment by Jason2866** [comment]
You find it in my unofficial builds https://jason2866.github.io/Tasmota-specials/
Look for LVGL

**Reply by booroondook-eng** [reply]
Thank you very much!
P.S. But now I have black screen on my Air101-LCD. When I put display.ini file into board's filesystem I have black screen too.
But when I change template to... 
`{"NAME":"LuatosC3","GPIO":[1,6210,736,704,1,1,800,768,1,1,8736,0,544,1,0,0,0,0,1,1,1,1],"FLAG":0,"BASE":1}`
(as described in the firtst message of this thread), then immediately after reboot I see the text "ST7735S" on a white background and this view blinks three times and then goes out and I see a black screen again.
Also, now I can't go into the Consoles menu because it causes a reboot.
To avoid this, I select the ESP32C3(1) template in the configuration menu. Then I can access the console again, but my display goes permanently black again.

**Reply by sfromis** [reply]
Check that your power source is sufficient, otherwise, you have a cause for such issues.

**Comment by Jason2866** [comment]
Try with: `:B,60,0` this disables DMA.

**Reply by booroondook-eng** [reply]
Thanks!!!

**Reply by barbudor** [reply]
I also had to disable DMA, otherwise the C3 was rebooting everytime I was getting on the "Consoles" menu
Any other web page is ok, only `/mn?` makes the C3 to reboot

I still have those bad pixels on the bottom line and the rightmost column
Also the curve seem to be inverted 
`Displaytext [B0C65535z]Hello` writes black on white
`Displaytext [B65535C0z]Hello` writes white on black
And colors are swapped: BGR instead of RGB

I tried to reduce the SPI clock from 16 to 8MHz but it didn't changed anything for the bad pixels
I can't seem to find any settings in display.ini about the curve and the RGB order

Any hints ?

![image](https://github.com/arendst/Tasmota/assets/2996042/0577ef3a-582a-499b-890a-4d6bb457e24e)

![image](https://github.com/arendst/Tasmota/assets/2996042/01121d6d-5aaf-472b-b22e-95e690415c91)

**Comment by xostia** [comment]
Not all the Air101-LCD are the same. If you have the one with the black (in the top) joystick, the LCD itself is different (in my case, I have both, and the one with the black dot in the joystick has no text about the model on the board). For the black one, check this display.ini

The RGB is inverted in this other model... and it seems to have a bit higher resolution.

:H,ST7735S,161,83,16,SPI,1,*,*,*,*,*,*,*,16
:S,-1,1,1,0,0,0
:I
01,80
11,E0
B1,3,02,35,36
B2,3,02,35,36
B3,6,02,35,36,02,35,36
B4,1,07
B6,2,B4,F0
C0,3,A2,02,84
C1,1,C5
C2,2,0A,00
C3,2,8A,2A
C4,2,8A,EE
C5,1,06
36,1,C8
E0,10,12,1C,10,18,33,2C,25,28,28,27,2F,3C,00,03,03,10
E1,10,12,1C,10,18,2D,28,23,28,28,26,2F,3B,00,03,03,10
F0,1,01
F6,1,00
3A,1,05
2A,4,00,02,00,81
2B,4,00,03,00,82
29,E0
:o,28
:O,29
:A,2A,2B,2C,16
:R,36
:0,68,00,18,00
:1,CC,18,00,01
:2,A8,00,18,02
:3,08,18,00,03
:i,21,20
:B,60,0
# :i,20,21 (and 161,83 first line for joystick white) :i,21,20 (and 160,80 first line for joystick black)

**Comment by xostia** [comment]
Another detail about the Air101 LCDs. Theorically GPIO11 is to switch on and off the backlight but GPIO11 is not configurable from Tasmota (because GPIO11 is SPI_VCC and it powers the flash chip?). I was testing and connecting the "BL" pin to other free GPIO of the ESP32, you can switch on and off the light of the screen. Even defining the GPIO TEMPLATE in the code, Tasmota does not care about the configuration of GPIO11 (with GPIO 255 you can see that).

If someone can clarify/explain this better, I would thank you.

**Reply by Jason2866** [reply]
to switch GPIO 11 to an useable GPIO an efuse has to be burnt.
Info https://www.esp32.com/viewtopic.php?t=25906
So it needs checking how the flash is powered. If it is connected directly to 3.3V the fuse can be burnt and GPIO 11 can be used.
If the flash chip is powered via GPIO 11, do NOT burn the efuse. Board will stop working.

EDIT: Regarding Luatos C3 wiki, GPIO 11 can be used. From the wiki:

GPIO11 Unlock Instructions

ESP32C3 The default function of GPIO11(VDD_SPI) is to supply power to flash. VDD of Flash of this development board is directly connected to 3.3, so this IO can be used as GPIO.

The following is the operation flow. Note that the following operations can only be performed once and cannot be restored after changes (because fuse bits are set, not registers, one-time operation）

pip installation using python esptool `pip install esptool`
Insert the development board into the computer, you can see the port in the device manager, record the port number, for example COM20
Open a command line window and enter: `espefuse.py -p <port> burn_efuse VDD_SPI_AS_GPIO 1`
Look at the prompt and enter: `BURN`

**Reply by xostia** [reply]
Thank you. I burn the fuse but GPIO11 does not appear with command GPIO 255.

I would like to use to on/off the screen (it works if you make a bridge with a free gpio). I do not know how to enable in the Tasmota code that I am adapting.

If anybody knows, it will be welcome. Thank you.

**Reply by sfromis** [reply]
I'd not expect Tasmota to "know" what you did with the efuse, and thus it will still count the pin as being unavailable. From the block of gpio 11-17, only gpio 12,13 seems to be "allowed" as potentially available depending on flash type. Tasmota has hardcoded logic to determine which pins are "hidden", and which are only red-marked.
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_support/support.ino#L1678-L1713

The `FlashPin` function is also called from the `Gpio` command to determine what's can be used.

AFAICT, source code modification would be the only way to open up gpio 11, maybe red-marked like gpio 12/13. Doing this would likely break compatibility with existing templates, and maybe also gpio config, but it should be feasible to config from scratch, I think....

**Reply by xostia** [reply]
This file is a bit different in the 13.1 version. It is a bit crazy because in this one 11 is not rejected, but it seems that 12 and 13, yes... but I can use 12 and 13.

GPIO11 is reserved (backlight on/off) -> :-(
GPIO13 is "shared" between one of the leds and the Air101LCD -> :-(
GPIO12 works without problem

If Air101LCD and the ESP32-C3 CORE are "brothers" why they designed this so bad!!!!!?

The "RedPin" code is this one, no red for GPIO11. Thanks for your help, Sfromis.

bool RedPin(uint32_t pin) // pin may be dangerous to change, display in RED in template console
{
#if defined(ESP32) && CONFIG_IDF_TARGET_ESP32C3
  return (12==pin)||(13==pin);  // ESP32C3: GPIOs 12 13 are usually used for Flash (mode QIO/QOUT)
#elif defined(CONFIG_IDF_TARGET_ESP32S2)
  return false;     // no red pin on ESP32S3
#elif defined(CONFIG_IDF_TARGET_ESP32S3)
  return (33<=pin) && (37>=pin);  // ESP32S3: GPIOs 33..37 are usually used for PSRAM
#elif defined(CONFIG_IDF_TARGET_ESP32)  // red pins are 6-11 for original ESP32, other models like PICO are not impacted if flash pins are condfigured
  // PICO can also have 16/17/18/23 not available
  return ((6<=pin) && (11>=pin)) || (16==pin) || (17==pin);  // TODO adapt depending on the exact type of ESP32
#else // ESP8266
  return (9==pin)||(10==pin);
#endif
}

**Reply by sfromis** [reply]
Of course, I looked at current source, where 12 and 13 are **not** "rejected", as they are specifically only red-marked, and therefore available. To make the "normally unavailable" gpio 11 "visible", it would make sense to have it red-marked, instead of looking like a normal gpio.

**Reply by xostia** [reply]
I found it! It is this function:

bool FlashPin(uint32_t pin)
{
#if defined(ESP32) && CONFIG_IDF_TARGET_ESP32C3
  return (((pin > 11) && (pin < 11)) || ((pin > 13) && (pin < 18))); 

I removed the first part of the IF (11 is allowed) and now in console I can see GPIO11 with "GPIO 255"

"GPIO11":{"0":"None"}

... I must check, because GPIO11 is defined in user_config_override.h as 922 (blacklight) but it appears as "0"

EDIT: I reset the device in the web interface and now it appears as

{"992":"Backlight"}

But I do not know how to ON/OFF backlight ... jajaja (probably it will work if I define it as one of the Leds)

Thank you, Sfromis

**Reply by sfromis** [reply]
Assuming that you're talking about using gpio 11 as a display backlight pin, turning it on/off would normally be on the "new" Power channel using the Power command or Tasmota front page toggle.

**Reply by xostia** [reply]
Yes, perfect... before this, the ON/OFF only left screen BLACK but now it switches off backlight. Fantastic!!

I do not know if somebody can update the information about this ESP32C3 CORE and the LCD in Tasmota web... it can be useful for somebody.

I have a sli

**Reply by sfromis** [reply]
The usual way for particular devices would be to create a template page like this:
https://templates.blakadder.com/luatos_CORE-ESP32.html
Having both the needed template, and any special procedure needed for setup. As you needed source modifications to get it to work, it may be "a bit too much" to guide users through on such a page, and it might make sense to have a build time options to control the tweaks for this device. I'm not involved in this, and I'm not sure what the "right" approach would be there.

**Reply by Jason2866** [reply]
@arendst Can you change for the C3 the Gpio11 pin available as red labeled?

**Reply by arendst** [reply]
I will.

**Reply by arendst** [reply]
Done. See latest dev branch

**Reply by vvvyky** [reply]
> 完成了。查看最新开发分支



> 完成了。查看最新开发分支

How to fix C3 using Gpio11？

**Reply by Jason2866** [reply]
@vvvyky Do as described https://github.com/arendst/Tasmota/discussions/18350#discussioncomment-8199939
